SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyRefCasts.swift

```diff
@@ -86,6 +88,15 @@ private extension UnaryInstruction {
       operand.set(to: replacement, context)
     }
 
+    if let ccb = self as? CheckedCastBranchInst {
+        // Make sure that loosing the formal source type is harmless (i.e the
+        // formal source type lowered equaled the operand SIL type before
+        // transformation).
+        let formalSrcType =  ccb.formalSrcType
+        if formalSrcType.loweredType(in: parentFunction, maximallyAbstracted: false) == oldOperandType {
```

Comment:
Why this weird check?

https://github.com/swiftlang/swift/pull/67583

It seems we can't always use the operands lowered SIL type (it looses some information?!) but need to carry the formal type.

So this check here checks that the we we lower the formal type to a source type we get the same type ie. we are not one of the cases where during the process of SIL lowering information was lost.

Reply:
Except, that as I write this I realize the check should be `formalSrcType == oldOperandType.astType`